jax/experimental/sparse/bcoo.py

```diff
@@ -69,25 +71,26 @@ def _bcoo_nse(mat, n_batch=0, n_dense=0):
   mask = mask.sum(list(range(n_batch, mask.ndim)))
   return mask.max()
 
-def _dedupe_bcoo(data, indices, shape):
+def _bcoo_sum_duplicates(data, indices, shape, nse=None):
   props = _validate_bcoo(data, indices, shape)
-  if indices.shape[:props.n_batch] != data.shape[:props.n_batch]:
-    # TODO: handle broadcasted dimensions.
-    raise NotImplementedError("dedupe_bcoo for broadcasted dimensions.")
-  f = functools.partial(_dedupe_bcoo_one,
-      shape=shape[props.n_batch:props.n_batch + props.n_sparse])
+  f = functools.partial(_bcoo_sum_duplicates_unbatched,
+      shape=shape[props.n_batch:props.n_batch + props.n_sparse], nse=nse)
   for _ in range(props.n_batch):
-    f = vmap(f)
+    f = broadcasting_vmap(f)
   return f(data, indices)
 
-def _dedupe_bcoo_one(data, indices, *, shape):
-  nse, = data.shape
-  assert indices.shape == (nse, len(shape))
+def _bcoo_sum_duplicates_unbatched(data, indices, *, shape, nse):
+  assert indices.shape == (data.shape[0], len(shape))
   if indices.shape[1] == 0:
     return data, indices
-  indices_unique, inv_idx = jnp.unique(indices, axis=0, return_inverse=True,
-                                       size=nse, fill_value=jnp.array(shape))
-  data_unique = jnp.zeros_like(data).at[inv_idx].add(data)
+  try:
+    indices_unique, inv_idx = jnp.unique(indices, axis=0, return_inverse=True,
+                                         size=nse, fill_value=jnp.array(shape))
+  except NonConcreteBooleanIndexError:
+    raise ValueError("In a traced context, calling BCOO.sum_duplicates() "
+                     "requires passing a non-None value for the nse argument.")
```

Comment:
Echoing @mattjj's comment in https://github.com/google/jax/pull/8305/files#r733105418, this is another case where ideally we would avoid referring to tracing, although I'm not sure what to say instead.

Reply:
Maybe "When used with JIT, vmap, or another transform..."